logstash-core/src/main/java/org/logstash/config/source/Base.java

```diff
@@ -0,0 +1,145 @@
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *	http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.logstash.config.source;
+
+import org.jruby.RubyArray;
+import org.jruby.RubyObject;
+import org.jruby.RubyString;
+import org.jruby.runtime.builtin.IRubyObject;
+import org.logstash.config.ir.PipelineConfig;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.logstash.RubyUtil.RUBY;
+
+public class Base {
+    protected RubyObject settings;
+    protected List<String> conflictMessages = new ArrayList<>();
+
+    public Base(RubyObject logstashSettings) {
+        settings = logstashSettings;
+    }
+
+    public List<String> getConflictMessages() {
+        return conflictMessages;
+    }
+
+    // this must be used when in Ruby code was @settings = <new value>
+    public void updateSettings(RubyObject settings) {
+        this.settings = settings;
+    }
+
+    public List<PipelineConfig> pipelineConfigs() {
+        throw new UnsupportedOperationException("`pipelineConfigs` must be implemented!");
+    }
+
+    public boolean isMatch() {
+        throw new UnsupportedOperationException("`isMatch` must be implemented!");
+    }
+
+    public boolean isConfigConflict() {
+        throw new UnsupportedOperationException("`isConfigConflict` must be implemented!");
+    }
+
+    //return subclass of LogStash::Setting
+    public IRubyObject configReloadAutomaticSetting() {
+        return this.settings.callMethod(RUBY.getCurrentContext(), "get_setting",
+                RubyString.newString(RUBY, "config.reload.automatic"));
+    }
+
+    public boolean configReloadAutomatic() {
+        final IRubyObject value = configReloadAutomaticSetting().callMethod(RUBY.getCurrentContext(), "value");
+        return value.toJava(Boolean.class);
+    }
+
+    public boolean isConfigReloadAutomatic() {
+        final IRubyObject valueIsSet = configReloadAutomaticSetting().callMethod(RUBY.getCurrentContext(), "set?");
+        return valueIsSet.toJava(Boolean.class);
+    }
+
+    //return subclass of LogStash::Setting
+    public IRubyObject configStringSetting() {
+        return this.settings.callMethod(RUBY.getCurrentContext(), "get_setting",
+                RubyString.newString(RUBY, "config.string"));
+    }
+
+    public String configString() {
```

Comment:
... these parts seem a little 'crazy' compared to the Ruby code in terms of what they do:
1. callMethod with dynamic Java -> Ruby String (`RubyString.newString(RUBY, "config.string")`)
2. callMethod("value") and convert to Java String
3. since this seems to only be called from Ruby -> convert Java String back to Ruby String

Reply:
if I want Ito call the Ruby method `get_setting("config.string")` from Java code to retrieve the setting was the only path it come in mind, the `callMethod` paramas are a String with method name and a list IRubyObject (the ruby method's parameter). If I want to retrieve the Setting named `config.string` from the Settings Ruby object is a there a better way to do?

Reply:
there likely is not (except a one where we would use a JRuby internal call-site). 
the point is (as mentioned later) that on its own these changes look like a step "backward" in terms of having additional overhead and reduced readability. it's easy to miss the big picture of the rewrite with these small PRs :smiling_imp: 
 
I think a bigger PR would have been justified here but I am fine with leaving this as is and follow up later on. 

still makes me wonder, from a higher level, how we should be addressing these rewrites. 
a good rule whether a Java rewrite is good-enough would be the amount o `callMethod` introduced :stuck_out_tongue_winking_eye: ?
on the other hand doing the changes as a direct rewrite little-by-little is straightforward but requires a refactoring follow up later so in the mean time it's going to be in a "Java that mostly does dynamic Ruby state" ...

Reply:
just thinking out loud, feel free to proceed to the other parts of the review ...

Reply:
I agree with your thoughts, `callMethod` pollute readability and make harder to understand the control flow. So generally to me seems a good metric. The problem here is that we have this dependency on `Settings`, I think that we could move upfront the Settings class (and the depending subclasses of Setting). Given that I'm in favor to hold on this PR and move the Settings before. Wha do you think @jsvd and @kares ?